fix: Use snprintf over sprintf for security reasons

matthewfeickert opened the pull request:
* Avoids:

```
 warning: 'sprintf' is deprecated: This function is provided for compatibility
 reasons only. Due to security concerns inherent in the design of sprintf(3),
 it is highly recommended that you use snprintf(3) instead.
```

* Use `sizeof` to determine the buffer size.

matthewfeickert commented:
@selvaggi @pavel-demin this is ready for review. Let me know if you have any questions.

pavel-demin commented:
I do not like the idea of ​​fixing something that is not broken.

If I am not mistaken, `sprintf` is specified in several versions of the C standard and is not deprecated in any version of the C standard. So I think that it is the warning message that should be fixed.

Additionally, some of these changes actually seem to break the working code:
- I do not think that `snprintf` will work in `awk` scripts
- [line 189 in external/tcl/tclCmdAH.c](https://github.com/delphes/delphes/blob/0f1ccd70a601816e0b6f6876271aa0ec76480aa4/external/tcl/tclCmdAH.c#L189) looks suspicious

matthewfeickert commented on external/tcl/tclCmdAH.c:
```diff
@@ -186,7 +186,7 @@ Tcl_CaseObjCmd(dummy, interp, objc, objv)
 	    
 	    arg = Tcl_GetStringFromObj(armPtr, &argLen);
-	    sprintf(msg, "\n    (\"%.*s\" arm line %d)", argLen, arg,
+	    snprintf(msg, sizeof(msg), sizeof(msg), "\n    (\"%.*s\" arm line %d)", argLen, arg,
```
```suggestion
	    snprintf(msg, sizeof(msg), "\n    (\"%.*s\" arm line %d)", argLen, arg,
```

> [line 189 in external/tcl/tclCmdAH.c](https://github.com/delphes/delphes/blob/0f1ccd70a601816e0b6f6876271aa0ec76480aa4/external/tcl/tclCmdAH.c#L189) looks suspicious

Yes, `git grep` error here.

matthewfeickert commented:
Aside: @pavel-demin please consider enabling GitHub Issues and Discussions. I would prefer to have discussed this in advance of making a PR and the current ticketing system https://cp3.irmp.ucl.ac.be/projects/delphes/report is not a good UX for anyone trying to find things from the past.

> If I am not mistaken, `sprintf` is specified in several versions of the C standard and is not deprecated in any version of the C standard.

Correct, though it is deprecated for all macOS as noted in https://debbugs.gnu.org/cgi/bugreport.cgi?bug=58966. Though it sounds like you are of a somewhat similar mind to Gnulib maintainers in https://mail.gnu.org/archive/html/bug-gnulib/2022-11/msg00013.html in that you'd prefer to not touch it. If that's the case, please feel free to close this PR.

> So I think that it is the warning message that should be fixed.

I do not think that Apple is going to change their compilers.

> I do not think that `snprintf` will work in `awk` scripts

Do you have tests that a user can run to verify changes? The CI only checks if things build, but is there a way for people who want to contribute to verify changes through a testing framework in advance?

pavel-demin commented:
I had actually been thinking about enabling issues for a few weeks. I have just enabled them.

pavel-demin commented:
I took another look at the changes that you are proposing to make and found other cases where the changes break the code instead of fixing it.

In Tcl code, there are multiple calls to `sprintf` where a pointer and not an array is used as the first argument to `sprintf`. In such cases, `sizeof` returns the size of the pointer and not the size of the array that the pointer points to.

Here is a minimal code example to illustrate the problem:
```
#include <stdio.h>

int main()
{
  char staticBuf[100];
  char *dst = staticBuf;
  printf("%d\n", sizeof(dst));
  return 0;
}
```

Here are some of the changes where this problem is present:
- [tclCmdAH.c#L1153-L1161](https://github.com/delphes/delphes/blob/78a67b83e821166145055652cf300dee07658e75/external/tcl/tclCmdAH.c#L1153-L1161)
- [tclHash.c#L335-L346](https://github.com/delphes/delphes/blob/78a67b83e821166145055652cf300dee07658e75/external/tcl/tclHash.c#L335-L346)
- [tclUtil.c#L2254](https://github.com/delphes/delphes/blob/78a67b83e821166145055652cf300dee07658e75/external/tcl/tclUtil.c#L2254)

Given the number of changes that need to be carefully reviewed and the number of problems that I have already found, I am afraid that I will not be able to fully review this pull request to the level where I am confident that these changes do not introduce new bugs.

matthewfeickert commented:
> I had actually been thinking about enabling issues for a few weeks. I have just enabled them.

That's great and thanks!

> Given the number of changes that need to be carefully reviewed and the number of problems that I have already found, I am afraid that I will not be able to fully review this pull request to the level where I am confident that these changes do not introduce new bugs.

That's fine. I'll put this into draft mode for the time being.

Do you have specific recommendations on additions to the CI for testing the full code beyond the build? If you enable the CI on this PR you'll note that the build passes, so it seems like it would be useful to have tests that can validate code changes.

pavel-demin commented:
After some thought and reading the documentation, I found out that these errors are actually caught by the compiler. They are warnings by default, but they can be converted to errors using the `-Werror=sizeof-pointer-memaccess` command line option.

matthewfeickert commented:
> Most of the errors introduced by this pull request are introduced in the Tcl code. Since Tcl is an external dependency to Delphes, I think that validating Tcl code changes is out of scope of the Delphes project.

Sure. I've rebased this (long delay was given travel and a conference) and dropped changes to externals as those can reasonably be upstreamed. I've also dropped changes to docs, given you mentioned this broke your current workflow.

The remaining change is the only to `examples/ExampleClCount.C`, not to the core source code, so I'll defer to you if you think that is even of use or not.

> After some thought and reading the documentation, I found out that these errors are actually caught by the compiler. They are warnings by default, but they can be converted to errors using the `-Werror=sizeof-pointer-memaccess` command line option.

Yes, sorry I should have mentioned that earlier but I didn't think it was of particular relevance as you're not compiling with that in your current workflow.

pavel-demin commented:
Thank you for updating this pull request. I do not think that the remaining change is really necessary since it is in an example script.

I am closing this pull request without merging it but will keep in mind the potential deprecation of the `sprintf` function. Hopefully this will never happen because, as this pull request shows, it is not trivial to fix it in Tcl code.

The other positive outcome of this pull request is that it motivated me to check other warnings and fix some of them.
